### PR TITLE
[Bugfix] Apply kv_b_proj LoRA across MLA attention paths

### LINKED ISSUE #48974: [Bug]: LoRA on MLA kv_b_proj is silently never applied — decode bypasses it via absorption, prefill via a stale pre-wrap reference
### Your current environment

vLLM 0.25.1 (V1), reproduced conceptually on current `main` (file refs below are `main`). Applies to any MLA model (DeepSeek V2/V3/R1/V3.2, GLM MoE DSA, Kimi K2) with an adapter targeting `kv_b_proj`.

### 🐛 Describe the bug

An adapter containing `kv_b_proj` LoRA tensors loads without error and is applied **nowhere**. The LoRA manager wraps the layer, but the wrapper's `forward()` is never invoked on any path:

- **Decode (absorbed/MQA):** attention consumes `W_UK`/`W_UV`, split from the raw `kv_b_proj` weight at load time, via BMMs (`vllm/model_executor/layers/attention/mla_attention.py:822`, `:1075`). No LoRA indirection exists in this path.
- **Prefill (MHA):** the up-projection calls `self.kv_b_proj(kv_c_normed)` on the attention *impl* (`mla_attention.py:2182`, `:2335`, `:2400`), which captured the raw `ColumnParallelLinear` at construction (`:2074`) — before LoRA wrapping. The impl is not an `nn.Module` (`vllm/v1/attention/backend.py:778`), so the LoRA manager's module-tree walk (`vllm/lora/model_manager.py:402`) rewires `MLAAttention.kv_b_proj` but never the impl's reference.

No warning, no error; all outputs and logprobs silently exclude the trained delta.

**Impact measured:** online-RL (trainer + vLLM sampler with per-step LoRA weight sync), rank-32 adapter on a 745B MLA MoE: trainer↔sampler KL grows from ~0.003 (numerics floor) to **0.035 within six optimizer steps**; module-group bisection at a frozen adapter attributes ~90% to `kv_b_proj` (symmetric removal: 0.035 → 0.0054). Likely the undiagnosed root cause of #17155.

**Repro:** serve any MLA model with an adapter whose `kv_b_proj` B is non-trivial; compare per-token logprobs against a merged-weights baseline — they diverge everywhere.

### Fix directions

1. **Decode:** apply the low-rank correction in absorbed mode — the delta factors onto the query and output sides (`q' += (q @ B_K) @ A`, `out += (attn @ A^T) @ B_V^T`), i.e. two shrink/expand pairs inside the MLA layer, no `B @ A` materialization. Same mechanism as SGLang's `srt/lora/deepseek_mla_correction.py`.
2. **Prefill:** the wrapped module must actually be invoked, with correct punica routing. Note rewiring `impl.kv_b_proj` alone is insufficient: vLLM slices decode+prefill rows from one scheduled batch, so a plain wrapped forward on `kv_c_normed[num_decode_tokens:]` would apply the mapping starting at row 0 (the decode tokens' adapter IDs), and the chunked-context calls up-project KV-cache tokens with no scheduler-row correspondence at all. (SGLang avoids this half structurally: its prefill call site is the LoRA-rewired attribute and extend/decode run as separate batches.)
3. **Short term / defense in depth:** fail closed — reject or warn on adapters targeting modules whose weights were absorbed, rather than silent partial application. An invariant that every accepted adapter tensor is applied somewhere would have caught this class of bug.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and checked the documentation and examples: closest is #17155 (stale-closed, undiagnosed — consistent with this root cause).

### PR DESCRIPTION
**Pull request by hershg**
Fixes #48974. MLA can bypass the wrapped `kv_b_proj` and use cached base projections, omitting its adapter. Apply token-routed LoRA corrections to materialized prefill and absorbed query/value projections, including sparse MLA.

Preserve local heads before DCP gathering and update AITER, AMX and Dots3Note callers. Round the final adapter delta to the output dtype before adding the base output, matching the native fallback and ordinary LoRA kernels; preserve the FP32 intermediate. Add an exact BF16 rounding-boundary regression and exercise LoRA/no-LoRA dispatch through the actual DCP attention entrypoint. Continues the closed, unmerged #49007.

## Testing

```bash
VLLM_TARGET_DEVICE=cpu PYTHONPATH=. uv run --no-project <cpu-env>/bin/python -m pytest tests/v1/attention/test_mla_context_chunks.py --confcutdir=tests/v1/attention -q --disable-warnings
```

At head `f6241b83857624a53e8d21ffd68bd119d357f8ef`, the CPU context suite passes 14 tests with 1 GPU-only skip, and applicable static/commit checks pass. The new cached-context regression uses the real chunk builder and prefill caller with nonzero request offsets; dropping the request slice fails its token-to-adapter mapping assertion. Both CPU DCP-dispatch cases also pass; restoring the old branch fails the LoRA case. The separately pinned v0.28 backport `fba3c24e` passed a GPU old-fails/new-passes boundary check and affected FP16/BF16 linear, query and value tests. Subsequent composed-runtime GLM PP1/PP2 synthetic parity results and exact settings are recorded in [SkyRL #2182](https://github.com/NovaSky-AI/SkyRL/pull/2182).

Those backport results do not qualify this main port. Current-head GPU/model validation remains outstanding, including adapter-versus-merged scores, chunked-prefix/mixed-adapter serving and backend coverage.

### CODE REVIEW

**Comment by hershg** [comment]
@MatthewBonanni @LucasWilkinson could you review the MLA LoRA fix? Rebased onto main `276db94f` at `12725c5`; the AMX weight scaling and earlier bot findings are addressed.

Local checks: 14 context tests passed (1 GPU-only skip), 2 DCP tests passed, and affected-file pre-commit passed. GPU validation of this main port is still outstanding; the v0.28 backport was tested separately.

GitHub pre-commit is skipped since I havent merged enough PRs yet). This continues #48974 / #49007.

File: vllm/models/dots3_note/nvidia/attention.py [hunk 1]
```diff
@@ -648,6 +657,12 @@ def forward_mha(
             kv_nope = self.kv_b_proj(kv_c)[0].view(
                 -1, self.num_heads, self.qk_nope_head_dim + self.v_head_dim
             )
+            apply_lora = getattr(kv_b_proj_lora, "apply_mla_kv_b_lora_linear", None)
+            if apply_lora is not None and request_lora_mapping is not None:
+                context_lora_mapping = request_lora_mapping[chunk.request_slice][
```
[the remark this answers is not in the document]
Comment (hershg) [reply vllm/models/dots3_note/nvidia/attention.py]: Fixed in 12725c5849244bd86b64a4b2898c27f8625463d5: the mapping slice uses chunk.req_start : chunk.req_end, matching _SlidingWindowChunk. Local pre-commit passed after rebase. A Dots3 GPU workload has not been tested.
